corehq/apps/sso/forms.py

```diff
@@ -595,11 +599,99 @@ class BaseSsoEnterpriseSettingsForm(forms.Form):
+            "This also removes the suggestion to copy an API key to another location when generating a new key."
+        ),
+        widget=BootstrapCheckboxInput(
+            inline_label=gettext_lazy(
+                "Always display full API keys to SSO users"
+            ),
+        ),
+    )
+
+    enforce_user_api_key_expiration = forms.BooleanField(
+        label=gettext_lazy("Require Expiration Date"),
+        required=False,
+        help_text=gettext_lazy(
+            "When enabled, this ensures that a date of expiration is always provided when SSO users "
+            "generated new keys from their API Keys page"
+        ),
+        widget=BootstrapCheckboxInput(
+            inline_label=gettext_lazy(
+                "Always require an expiration date for API keys"
+            ),
+        ),
+    )
+
+    max_days_until_user_api_key_expiration = forms.TypedChoiceField(
+        label=gettext_lazy("Maximum Expiration Length"),
+        required=False,
+        help_text=gettext_lazy(
+            "If an expiration date is required, the expiration date will not exceed this many days "
+            "from the date an API key is created"
+        ),
+        choices=(
+            ('', '-----'),
+            (365, gettext_lazy("1 Year")),
+            (180, gettext_lazy("180 Days")),
+            (160, gettext_lazy("160 Days")),
+            (120, gettext_lazy("120 Days")),
+            (90, gettext_lazy("90 Days")),
+            (60, gettext_lazy("60 Days")),
+            (30, gettext_lazy("30 Days")),
+        ),
+        coerce=int,
+        empty_value=None
+    )
+
+    expiration_warning_message = gettext_lazy(
+        'All existing API keys that do not comply with this expiration date policy will be updated to ensure a '
+        'compliant expiration date is present. For keys that do not have an expiration date set, the date of '
+        'expiration will be <span data-bind="text: expirationLength"></span> from today after clicking '
+        '"Update Configuration" below. This will apply only to SSO users associated with this Identity Provider.'
+    )
+
+    def __init__(self, identity_provider, *args, allow_multi_view_api_keys=False, **kwargs):
         self.idp = identity_provider
+        self.allow_multi_view_api_keys = allow_multi_view_api_keys
         super().__init__(*args, **kwargs)
 
     def get_primary_fields(self):
+        if self.allow_multi_view_api_keys:
+            fieldset = crispy.Fieldset(
+                _('API Key Management'),
+                checkbox_field('always_show_user_api_keys'),
+                checkbox_field('enforce_user_api_key_expiration', data_bind='checked: enforceExpiration'),
+                crispy.Div(
+                    crispy.Field(
+                        'max_days_until_user_api_key_expiration', data_bind="value: expirationLengthValue"),
+                    crispy.HTML(
```

Comment:
offline on this one too, but as a reminder: I think the best approach here is to use `render_to_string()` with a template loading `hq_shared_tags`. This can then reference the standard form offsets with `{% css_action_class %}` so that we are not hardcoding the offsets here—especially something to avoid in a python file always.

making a note to myself include examples of this in our styleguide

Reply:
Addressed in d36fdbacf03a01c74ccf7da2260370026db7fef9